[WIP][dashboard scoped filter] part 2: add algorithm to convert checked ids to scope object

graceguo-supercat opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
After filter scope selector UI, we will get filters scope data like this format:
```
{
    filterKey: '108___time_range',
    checkedChartIds: [101, 102, 103, 104],
}
```
dashboard has layout data in tree structure, a simplified example could be like:
```
{
    'ROOT_ID': { children: ['tab-A', 'tab-B'] }
    'tab-A': { children: [101, 102, 103]},
    'tab-B': { children: [104, 105]},
}
```
this PR will create an algorithm that creates _**filter scope**_ object for each filter field, and will be merged into dashboard redux state (as part of dashboardFilter state):
```
{
    18: {
        columns: {
            _time_range: ['last year'],
        },
        labels: {
            _time_range: "Time Range",
        },
        scopes: {
            _time_range: {
               scope: ['tab-A', 'tab-B'],
               immune: [105],
            },
        }
    }
}
```

**Basic rules to decide scope:**

- A `scope` object has 2 fields:
   - The minimum component unit for scope is `tab`.  A filter can apply to one or multiple tabs.
   - If a chart inside a tab but not applicable to a filter, it will be added to immune list.
```
{ 
    scope: [list of tab ids],
    immune: [list of chart ids],
}
```
- If any chart is applicable to a filter field, the chart's container tab should be in the filter's scope. Other charts inside the tab that are not applicable to the filter, will be added to immune list.
- If all sub tabs are applicable to a filter field, we mark the parent tab in the filter' scope. Any chart that under the parent tab and not applicable to the filter, will be added to immune list.


### TEST PLAN
new unit test is added.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue: #8404, #8522
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@etr2460 @mistercrunch

Commit:
filter scope selector modal

dpgaspar committed:
[docs] CTAS on PostgreSQL needs commit to apply (#8367)

* [docs] New, document need for PG to use autocommit for CTAS

Commit:
[Config] Cache static resources (#8370)

john-bodley committed:
[sqla] Enforcing ISO 8601 date/timestamp formats (#7702)

john-bodley committed:
[ci] Deprecate flake8 (#8409)

* [ci] Deprecate flake8

* Addressing @villebro's comments

francishemingway committed:
fixed typo in installation instructions (#8413)

MarcusSorealheis committed:
[tests] Fix, removing deprecated function for future Python release compatibility. (#8353)

* removing deprecated function for future Python release compatibility.

* also removed assertNotEquals per @dpgaspar

betodealmeida committed:
Add improved typed casting to BigQuery (#8331)

mistercrunch committed:
chore: disable another flaky cypress test (#8423)

Also fix isort issue on master

john-bodley committed:
[pydruid] Bumping the pydruid version (#8430)

mistercrunch committed:
docs: add a note to RELEASING about Slack Channel (#8419)

mistercrunch committed:
docs: Add an 'Orientation' section to CONTRIBUTING.md (#8422)

villebro committed:
[fix] refine release instructions (#8381)

* Separate RC from VERSION

* Fix pypi_push.sh and refine instructions

* Add SUPERSET_ prefix to env variables

* Finalize release instructions

* Change one-off to First Time Only

* Add tagging of final version

* Convert some remaining env variables and add a check step to pypi deploy

john-bodley committed:
[fix] Updating parse_human_timedelta typing (#8436)

austinpray committed:
spelling fix (#8442)

christophlingg committed:
explain the need to enable async queries (#8444)

christophlingg committed:
introduce a space in command line option (#8438)

see https://docs.celeryproject.org/en/latest/userguide/optimizing.html

Commit:
Split up tests/db_engine_test.py (#8449)

* Split up db_engine_specs_test.py into a number of targeted files

* Remove db_engine_specs_test.py

* isort

dpgaspar committed:
[setup] Fix, download_url (#8439)

* [setup] Fix, download_url

betodealmeida committed:
Allow fetching all rows from results endpoint (#8389)

* Allow bypassing DISPLAY_MAX_ROW

* Add unit tests and docs

* Fix tests

* Fix mock

* Fix unit test

* Revert config change after test

* Change behavior

* Address comments

Commit:
add single-field-edit in multi-edit mode switch

alexandrebrilhante committed:
removed repeated word (#8458)

dpgaspar committed:
[docs] Crafting a new release (#8433)

* [docs] Crafting a new release

Commit:
Add UI-only database configuration method for extended authorization scenarios (#8448)

* Add encrypted_extra to dbs

* WIP - UI-based BigQuery connection configuration

* Fix 500 bubbling to the surface when adding a database connection

* Add check for valid json

* black formatting:

* isort

* Incorporate PR feedback. Thanks all!

* black

* Typo fix in CONTRIBUTING.md

dpgaspar committed:
[db engine] Add support for Elasticsearch (#8441)

* [db engine] Add support for Elasticsearch

john-bodley committed:
[SIP-15] Making client time use UTC as the local time (#8450)

* [fix] Making client time UTC

* Update UPDATING.md

john-bodley committed:
[SIP-15] Adding initial framework (#8398)

* [sip-15] Adding initial framework

* [toast] Addressing etr2460's comments

* [fix] Addressing etr2460's comments

Commit:
fix code review comments (round 1)

Commit:
Provide documentation for using a Service Account to connect to BigQuery (#8462)

* Provide documentation for using a Service Account to connect to BigQuery

* Alter line wrapping for shorter lines

* Whitespace commit to trigger another build (flake)

* Another meaningless whitespace change to trigger another build

Commit:
[Datasource Editor] A few small UI changes in modal to prevent accidental edits (#8471)

john-bodley committed:
Update UPDATING.md (#8465)

john-bodley committed:
[sql] Fixing datetime SQL literal (#8464)

mistercrunch committed:
docs: more release instructions (#8432)

khtruong committed:
feat: add expand_data parameter (#8472)

* feat: add expand_data parameter

* fix: reformat files

Commit:
Use `config[]` not `config.get()` (#8454)

* Typo fix in CONTRIBUTING.md

* Alter references to config.get('FOO') to use preferred config['FOO']

* Set missing configuration constants in superset/config.py

* Misc. CI fixes

* Add type annotation for FEATURE_FLATGS

Commit:
Re-enable DRUID_IS_ACTIVE flag (#8482)

* Re-enable DRUID_IS_ACTIVE flag

* Make CI happy

MarcusSorealheis committed:
fixing typo. (#8487)

christophlingg committed:
Enhance docs for email reporting (#8486)

* extend documentation about setting up email reporting

* mention EMAIL_REPORTS_USER

john-bodley committed:
[sip-15] Fix time range endpoints encoding (#8481)

john-bodley committed:
[sip-15] Adding database level python-date-format (#8474)

conglei committed:
feat(TableVis): Switch to the new table plugin (#8446)

* use new table comp

* remove commented code

* update lock file

* fix lint

* bump version

* update lock file

* fix lock issue

Commit:
[datasource editor] Only one click target for edit action (#8495)

john-bodley committed:
[SIP-15] Updating ISO 8601 regex match to make time optional (#8494)

francishemingway committed:
Change Fedora installation instructions + some small formatting changes (#8496)

yahwang committed:
feat: add korea(South) info on CountryMap viz and bump legacy-plugin-chart-country-map to v0.11.4 (#8478)

paulvic committed:
Add `Duplicate Tab` option (#8485)

* Add `Duplicate Tab` option

Adds an option to duplicate the current tab's setting/content into a new tab. Useful if you're iterating on a query.

* Add test

MarcusSorealheis committed:
build: bump dompurify version because of nasty xss bypass. (#8498)

serenajiang committed:
[typing] superset/models/core.py (#8284)

betodealmeida committed:
Allow user to customize query cost estimate (#8470)

* Allow user to customize query estimate

* Add docs; run black

* Update docs with types

john-bodley committed:
[SIP-15] Adding grace period (#8490)

liuyanpunk committed:
Updating pip with default timeout (#8492)

* Updating pip with default timeout

Most of the time, the pip default site is realiable but we do see timeouts pops out when network is busy or temporary unavaliable .

The default value was 15 and that's way too short. So adding --default-time and set it to 1000 should be long enough to mitigate

* Update Dockerfile

Commit:
Fix column type on dbs.encrypted_extra, add instructions for testing migration downgrades (#8493)

* Fix column type on dbs.encrypted_extra

* Add instructions for testing migration downgrades

* Account for non-Postgres DBs in migration

* Use batch_alter_table to make SQLite happy

* Another CI-appeasing tweak

paulvic committed:
Update query tab's title when saving a new query (#8484)

* Update query tab's title when saving a new query

Just like it does when you opt to update an existing query

* Remove trailing whitespace

dpgaspar committed:
[docs] Small fixes on the release docs (#8500)

* [docs] Small fixes on the release docs

* [docs] Fix, remove deprecated app parameter from fab create-admin

john-bodley committed:
[SIP-15] Making sure the toast link opens in a new tab (#8503)

* [SIP-15] Ensuring toast opens in a new tab

* Update config.py

betodealmeida committed:
Remove tabs (#8508)

Commit:
refactory after design review

Commit:
Make Report a bug link open in a new tab (#8509)

conglei committed:
revert tablevis (#8515)

john-bodley committed:
[sip-15] Fixing time range endpoints from dashboards (#8513)

Zanoellia committed:
Default page: remove title attribute to fall back on appbuilder.app_name (#8427)

villebro committed:
Replace hardcoded default colors with defaults in setupColors.js (#8510)

villebro committed:
[fix] Improve csv upload functionality (#8457)

* [fix] csv upload when table metadata present

* Remove table from hive spec

* Move upload before table metadata creation

* Refine upload logic, dd unit tests and fix translations

* Use ALLOWED_EXTENSIONS from config

* Address review comments

* Fix error message grammar

* Add return type to hive csv upload and replace first with one_or_none

etr2460 commented on superset/assets/src/dashboard/util/getFilterScopeNodesTree.js:
```diff
@@ -25,83 +27,93 @@ import {
-      };
-    }
+  const type = currentNode.type;
+  if (CHART_TYPE === type && currentNode.meta.chartId) {
```
this will throw an error if currentNode isn't defined (as `{}.meta.chartId` will throw)

graceguo-supercat replied:
added extra check.

etr2460 commented on superset/assets/src/dashboard/util/getSelectedChartIdForFilterScopeTree.js:
```diff
@@ -0,0 +1,53 @@
+          getChartIdAndColumnFromFilterKey(filterKey).chartId !== chartId,
+      )
+    ) {
+      return 0;
```
same comment as previous PR about returning null or a const here

graceguo-supercat replied:
this return null is fixed in #8404.

etr2460 reviewed (commented):
first pass!

john-bodley committed:
[docs] Adding Slice parameter documentation (#8531)

* [CONTRIBUTING.md] Adding Slice parameter documentation

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

francishemingway committed:
[docs] User tutorial (#8459)

* added user tutorial and rearranged docs hierarchy

* added license header, renamed admin tutorial file

* fixed image url issue

* Fix spelling

* Tweaks to text based on feedback

* guilabel and menu test

* added guilabel and upload csv section

* tidy up of rst formatting

dpgaspar committed:
[cli] Fix, import datasources exported by UI (#8477)

* [cli] Fix, import datasources exported by UI

john-bodley committed:
[docs] Updating slice parameter documentation. (#8535)

* Update CONTRIBUTING.md

* Update CONTRIBUTING.md

Commit:
fix a few props initial value

graceguo-supercat commented:
I just realized I messed up this PR with some commits in another PR. so i rebased it again. This PR should only contain changes for the function that converts checked chart ids to scope object (and its unit test).

For the comments about filter scope selector, I fixed them in #8404.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/FilterScopeSelector.jsx:
```diff
@@ -403,100 +418,85 @@ export default class FilterScopeSelector extends React.PureComponent {
+    return (
+      <div className="selected-fields multi-edit-mode">
+        {currentFilterLabels.length === 0 && t('No filter is selected.')}
+        {currentFilterLabels.length === 1 && t('Editing 1 filter: ')}
```
i removed trilling space. The translation function allow simple template: 
`t('Batch editing %d filters:', currentFilterLabels.length)`
I fixed it in #8404.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/renderFilterFieldTreeNodes.jsx:
```diff
@@ -25,7 +25,9 @@ export default function renderFilterFieldTreeNodes({
 }) {
-  return nodes.map(node => ({
+  const root = nodes[0];
+  const allFilterNodes = nodes[0].children;
```
fixed.

graceguo-supercat commented on superset/assets/src/dashboard/components/filterscope/renderFilterFieldTreeNodes.jsx:
```diff
@@ -25,7 +25,9 @@ export default function renderFilterFieldTreeNodes({
 }) {
-  return nodes.map(node => ({
+  const root = nodes[0];
+  const allFilterNodes = nodes[0].children;
```
fixed.

graceguo-supercat commented on superset/assets/spec/javascripts/dashboard/util/getFilterScopeFromNodesTree_spec.js:
```diff
@@ -0,0 +1,216 @@
+      }),
+    ).toEqual({
+      scope: ['ROOT_ID'],
+      immune: [105],
```
yes. For this test case, the filter apply to all charts under `ROOT_ID`, which has charts [104, 105, 106, 108]. chart 108 is filter_box itself, it will be immune automatically.

nytai committed:
fab 2.2.0 (#8517)

paulvic committed:
Allow customization of documentation icon and text (#8507)

* Allow customization of documentation icon and text

* Set icon width to 100%

* Use double quotes for strings

axuew committed:
Add support for database engine SAP Hana (#8411)

* Add support for database engine SAP Hana

* Support hana services

Increase time, minute, and second

* Fix hana return string

* Fix formatting errors

dpgaspar committed:
[docs] 0.35.0 CHANGELOG (#8547)

* [docs] Update CHANGELOG with 0.35.0

* [docs] Fix, github md problem

Commit:
Upgrade superset-ui/time-format (#8541)

rusackas committed:
Clickable checkbox labels (#8534)

* passing onClick prop to header with the existing onChange method.

* basic test checking that label click fires the onChange method.

* cleaning up stuff caught by linting.

betodealmeida committed:
Encode feature flags to JSON pessimistically (#8529)

* Encode feature flags to JSON pessimistically

* Add unit test

* Remove old imports

john-bodley committed:
[fix] Re-cleanup legacy filters (#8523)

villebro committed:
Add convert_dttm method to SnowflakeEngineSpec (#8551)

Commit:
convert ids to scope object

graceguo-supercat commented:
please comment on #8559.
